Kernel/Time/TimeManagement.cpp

```diff
@@ -0,0 +1,249 @@
+            if (!probe_and_set_legacy_hardware_timers())
+                ASSERT_NOT_REACHED();
+        return;
+    }
+    ASSERT(probe_and_set_legacy_hardware_timers());
+}
+
+Vector<size_t> TimeManagement::scan_and_set_periodic_timers()
+{
+    bool enable_periodic_mode = is_hpet_periodic_mode_allowed();
+    dbg() << "Scanning for Periodic timers";
+    Vector<size_t> periodic_timers_indexes;
+    periodic_timers_indexes.ensure_capacity(m_hardware_timers.size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        if (!m_hardware_timers[index].is_null()) {
+            if (m_hardware_timers[index]->is_periodic_capable()) {
+                periodic_timers_indexes.append(index);
+                if (enable_periodic_mode)
+                    m_hardware_timers[index]->set_periodic();
+            }
+        }
+    }
+    return periodic_timers_indexes;
+}
+
+Vector<size_t> TimeManagement::scan_for_non_periodic_timers()
+{
+    dbg() << "Scanning for Non-Periodic timers";
+    Vector<size_t> non_periodic_timers_indexes;
+    non_periodic_timers_indexes.ensure_capacity(m_hardware_timers.size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        if (!m_hardware_timers[index].is_null())
+            if (!m_hardware_timers[index]->is_periodic_capable())
+                non_periodic_timers_indexes.append(index);
+    }
+    return non_periodic_timers_indexes;
+}
+
+bool TimeManagement::is_hpet_periodic_mode_allowed()
+{
+    if (!KParams::the().has("hpet")) {
+        return true;
+    }
+    auto hpet_mode = KParams::the().get("hpet");
+    if (hpet_mode == "periodic")
+        return true;
+    if (hpet_mode == "nonperiodic")
+        return false;
+    ASSERT_NOT_REACHED();
+}
+
+bool TimeManagement::probe_and_set_non_legacy_hardware_timers()
+{
+    if (!ACPI::Parser::the().is_operable())
+        return false;
+    if (!HPET::test_and_initialize())
+        return false;
+    if (!HPET::the().comparators().size()) {
+        dbg() << "HPET initialization aborted.";
+        return false;
+    }
+    dbg() << "HPET: Setting appropriate functions to timers.";
+
+    m_hardware_timers.resize(HPET::the().comparators().size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        m_hardware_timers[index] = HPET::the().comparators()[index];
+#ifdef TIME_DEBUG
+        dbg() << m_hardware_timers[index].ptr() << " <- " << HPET::the().comparators()[index].ptr();
+#endif
+    }
+
+    auto periodic_timer_indexes = scan_and_set_periodic_timers();
+    auto non_periodic_timer_indexes = scan_for_non_periodic_timers();
+
+    if (is_hpet_periodic_mode_allowed())
+        ASSERT(!periodic_timer_indexes.is_empty());
+
+    if (!non_periodic_timer_indexes.is_empty())
+        m_system_timer = m_hardware_timers[non_periodic_timer_indexes[0]];
+    else
+        m_system_timer = m_hardware_timers[periodic_timer_indexes[1]];
```

Comment:
Why is it safe to assume that `periodic_timre_indexes` has 2 or more entries here?